src/osd/PrimaryLogPG.cc

```diff
@@ -3395,38 +3420,51 @@ void PrimaryLogPG::dec_all_refcount_manifest(const object_info_t& oi, OpContext*
     ctx->register_on_commit(
       [oi, this](){
 	refcount_manifest(oi.soid, oi.manifest.redirect_target, 
-			  refcount_t::DECREMENT_REF, NULL);
+			  refcount_t::DECREMENT_REF, NULL, std::nullopt);
       });
   } 
 }
 
-void PrimaryLogPG::refcount_manifest(hobject_t src_soid, hobject_t tgt_soid, refcount_t type, 
-				     RefCountCallback* cb)
+ceph_tid_t PrimaryLogPG::refcount_manifest(hobject_t src_soid, hobject_t tgt_soid, refcount_t type,
+                                     Context *cb, std::optional<bufferlist> chunk)
 {
   unsigned flags = CEPH_OSD_FLAG_IGNORE_CACHE | CEPH_OSD_FLAG_IGNORE_OVERLAY |
-                   CEPH_OSD_FLAG_RWORDERED;                      
+                   CEPH_OSD_FLAG_RWORDERED;
+
+  dout(10) << __func__ << " Start refcount from " << src_soid
+           << " to " << tgt_soid << dendl;
 
-  dout(10) << __func__ << " Start refcount from " << src_soid 
-	   << " to " << tgt_soid << dendl;
-    
   ObjectOperation obj_op;
   bufferlist in;
-  if (type == refcount_t::INCREMENT_REF) {             
+  if (type == refcount_t::INCREMENT_REF) {
     cls_cas_chunk_get_ref_op call;
     call.source = src_soid.get_head();
-    ::encode(call, in);                             
+    ::encode(call, in);
     obj_op.call("cas", "chunk_get_ref", in);
-  } else if (type == refcount_t::DECREMENT_REF) {                    
+  } else if (type == refcount_t::DECREMENT_REF) {
     cls_cas_chunk_put_ref_op call;
     call.source = src_soid.get_head();
-    ::encode(call, in);          
+    ::encode(call, in);
     obj_op.call("cas", "chunk_put_ref", in);
-  }                                                     
-  
-  Context *c = nullptr;
+  } else if (type == refcount_t::CREATE_OR_GET_REF) {
+    cls_cas_chunk_create_or_get_ref_op get_call;
+    get_call.source = src_soid.get_head();
+    ceph_assert(chunk);
+    get_call.data = move(*chunk);
+    ::encode(get_call, in);
+    obj_op.call("cas", "chunk_create_or_get_ref", in);
+  } else {
+    ceph_assert(0 == "unrecognized type");
+  }
+
+  Context *c = nullptr, *fin = nullptr;
   if (cb) {
-    C_SetManifestRefCountDone *fin =
-      new C_SetManifestRefCountDone(cb, src_soid);
+    if (type == refcount_t::INCREMENT_REF ||
+        type == refcount_t::DECREMENT_REF) {
+      fin = new C_SetManifestRefCountDone(static_cast<RefCountCallback*>(cb), src_soid);
```

Comment:
This is a really confusing asymmetry for this helper.  You needed to do this because start_dedup issues multiple refcount_manifest calls on a single manifest_ops entry, so you need to assume that the caller has already handled manifest_ops and cb already takes care of it. It just so happens that start_dedup is the only user of CREATE_OR_GET_REF, so special casing it avoided the problem at the expense of making this helper's behavior randomly different between CREATE_OR_GET_REF and INCREMENT_REF.  DECREMENT_REF never passes cb, so that case is irrelevant here.  INCREMENT_REF happens in two places, I think you should should move the manifest_ops handling inline in those two places or create a helper to do so and make this method agnostic of what cb does.